dss-pades-openpdf/src/main/java/eu/europa/esig/dss/pdf/openpdf/ITextDocumentReader.java

```diff
@@ -104,13 +120,43 @@ public ITextDocumentReader(DSSDocument dssDocument) throws IOException, InvalidP
 	 * @throws IOException if an exception occurs
 	 * @throws InvalidPasswordException if the password is not provided or invalid for a protected document
 	 */
-	public ITextDocumentReader(DSSDocument dssDocument, byte[] passwordProtection) throws IOException, InvalidPasswordException {
+	public ITextDocumentReader(DSSDocument dssDocument, byte[] passwordProtection, PdfMemoryUsageSetting pdfMemoryUsageSetting)
+			throws IOException, InvalidPasswordException {
 		Objects.requireNonNull(dssDocument, "The document must be defined!");
+		Objects.requireNonNull(pdfMemoryUsageSetting, "PdfMemoryUsageSetting must be defined!");
 		this.dssDocument = dssDocument;
-		try (InputStream is = dssDocument.openStream()) {
-			this.pdfReader = new PdfReader(is, passwordProtection);
+		try {
+			if (PdfMemoryUsageSetting.Mode.MEMORY_FULL != pdfMemoryUsageSetting.getMode() && dssDocument instanceof FileDocument) {
+				FileDocument fileDocument = (FileDocument) dssDocument;
+				this.pdfReader = getFileDocumentPdfReader(fileDocument, passwordProtection, pdfMemoryUsageSetting);
+
+			} else if (dssDocument instanceof InMemoryDocument) {
+				InMemoryDocument inMemoryDocument = (InMemoryDocument) dssDocument;
+				this.pdfReader = new PdfReader(inMemoryDocument.getBytes(), passwordProtection);
+
+			} else {
+				try (InputStream is = dssDocument.openStream()) {
+					this.pdfReader = new PdfReader(is, passwordProtection);
+				}
+			}
+
 		} catch (BadPasswordException e) {
-            throw new InvalidPasswordException(String.format("Encrypted document : %s", e.getMessage()));
+			throw new InvalidPasswordException(String.format("Encrypted document : %s", e.getMessage()));
+		}
+	}
+
+	private PdfReader getFileDocumentPdfReader(FileDocument fileDocument, byte[] passwordProtection,
+											   PdfMemoryUsageSetting pdfMemoryUsageSetting) throws IOException {
+		String filenameSource = fileDocument.getFile().getAbsolutePath();
+		switch (pdfMemoryUsageSetting.getMode()) {
+			case MEMORY_BUFFERED:
+				// This condition uses a MappedByteBuffer to process the file in memory
+				return new PdfReader(filenameSource, passwordProtection);
+			case FILE:
+				return new PdfReader(new RandomAccessFileOrArray(filenameSource, false, true), passwordProtection);
```

Comment:
the `RandomAccessFileOrArray` is declared as Closable, neither the openpdf is actually closing it.. instead shall we close it here?

Reply:
Good catch. I will check if closing it here won't brake the processing. Otherwise, I will do a similar logic with PdfBoxDocumentReader and RandomAccessRead handling.

Reply:
I did a look in the code of OpenPdf and it looks like the `RandomAccessFileOrArray` is actually being closed inside of `PdfReader.close()` method. See that `RandomAccessFileOrArray` is stored inside `PRTokeniser` object (`tokens` variable) https://github.com/LibrePDF/OpenPDF/blob/1.3.43/openpdf/src/main/java/com/lowagie/text/pdf/PdfReader.java#L284, which is closed in `PdfReader.close()` https://github.com/LibrePDF/OpenPDF/blob/1.3.43/openpdf/src/main/java/com/lowagie/text/pdf/PdfReader.java#L3408 and https://github.com/LibrePDF/OpenPDF/blob/1.3.43/openpdf/src/main/java/com/lowagie/text/pdf/PRTokeniser.java#L182.

So I think in order to not complexify the code, I will keep it as it is.

Moreover, I will go even further and will align the PdfBox implementation accordingly, as `RandomAccessReadBuffer` is actually being closed as well inside the `PDDocument`. See https://github.com/apache/pdfbox/blob/3.0.3/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/PDDocument.java#L1272.